### ISSUE TITLE
E-Government SmartCard not forwarded properly

### ISSUE THREAD

**Issue by Raven24** [FreeRDP/FreeRDP#6420]
**Describe the bug**
I'm trying to forward a smart card (Austrian e-healt/e-government) using xfreerdp and a ReinerSCT device with pin pad.
The reader is detected by the CyberJack device manager and can be tested successfully (first screenshot), although there are some warnings in the output that don't appear when the device is directly attached (second screenshot).
When I try to use the A-Sign Client app it doesn't show any card inserted.
Please let me know, if you need more debugging or when there is a code branch I can test.

I've included the log where I connect to the remote end, run the device test once and then try to read the card in the client app.

**Log**
[freerdp_smartcard.txt](https://github.com/FreeRDP/FreeRDP/files/5033822/freerdp_smartcard.txt)

**Screenshots**
- RDP
![image](https://user-images.githubusercontent.com/117366/89509874-23ae0d00-d7d0-11ea-8c42-1d75e256a3ad.png)
![image](https://user-images.githubusercontent.com/117366/89514112-b4d3b280-d7d5-11ea-97a1-0dea548e67ac.png)

- Directly
![image](https://user-images.githubusercontent.com/117366/89513953-83f37d80-d7d5-11ea-9da8-e2cc2fcea6fe.png)
![image](https://user-images.githubusercontent.com/117366/89514531-46dbbb00-d7d6-11ea-838f-81aca501d604.png)

**Application details**
```
This is FreeRDP version 3.0.0-dev (d87d351c0)
Build configuration: BUILD_TESTING=OFF BUILTIN_CHANNELS=ON HAVE_AIO_H=1 HAVE_EXECINFO_H=1 HAVE_FCNTL_H=1 HAVE_GETLOGIN_R=1 HAVE_INTTYPES_H=1 HAVE_JOURNALD_H=TRUE HAVE_MATH_C99_LONG_DOUBLE=1 HAVE_POLL_H=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK=ON HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL= HAVE_SYSLOG_H=1 HAVE_SYS_EVENTFD_H=1 HAVE_SYS_FILIO_H= HAVE_SYS_MODEM_H= HAVE_SYS_SELECT_H=1 HAVE_SYS_SOCKIO_H= HAVE_SYS_STRTIO_H= HAVE_SYS_TIMERFD_H=1 HAVE_TM_GMTOFF=1 HAVE_UNISTD_H=1 HAVE_XI_TOUCH_CLASS=1 WITH_ALSA=ON WITH_CAIRO=OFF WITH_CCACHE=ON WITH_CHANNELS=ON WITH_CLANG_FORMAT=ON WITH_CLIENT=ON WITH_CLIENT_AVAILABLE=1 WITH_CLIENT_CHANNELS=ON WITH_CLIENT_CHANNELS_AVAILABLE=1 WITH_CLIENT_COMMON=ON WITH_CLIENT_INTERFACE=OFF WITH_CUPS=ON WITH_DEBUG_ALL=OFF WITH_DEBUG_CAPABILITIES=OFF WITH_DEBUG_CERTIFICATE=OFF WITH_DEBUG_CHANNELS=OFF WITH_DEBUG_CLIPRDR=OFF WITH_DEBUG_DVC=OFF WITH_DEBUG_KBD=OFF WITH_DEBUG_LICENSE=OFF WITH_DEBUG_MUTEX=OFF WITH_DEBUG_NEGO=OFF WITH_DEBUG_NLA=OFF WITH_DEBUG_NTLM=OFF WITH_DEBUG_RAIL=OFF WITH_DEBUG_RDP=OFF WITH_DEBUG_RDPDR=OFF WITH_DEBUG_RDPEI=OFF WITH_DEBUG_RDPGFX=OFF WITH_DEBUG_REDIR=OFF WITH_DEBUG_RFX=OFF WITH_DEBUG_RINGBUFFER=OFF WITH_DEBUG_SCARD=OFF WITH_DEBUG_SND=OFF WITH_DEBUG_SVC=OFF WITH_DEBUG_SYMBOLS=OFF WITH_DEBUG_THREADS=OFF WITH_DEBUG_TIMEZONE=OFF WITH_DEBUG_TRANSPORT=OFF WITH_DEBUG_TSG=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF_AVAILABLE=0 WITH_DEBUG_URBDRC=OFF WITH_DEBUG_WND=OFF WITH_DEBUG_X11=OFF WITH_DEBUG_X11_CLIPRDR=OFF WITH_DEBUG_X11_LOCAL_MOVESIZE=OFF WITH_DEBUG_XV=OFF WITH_DSP_EXPERIMENTAL=OFF WITH_DSP_FFMPEG=OFF WITH_EVENTFD_READ_WRITE=1 WITH_FAAC=OFF WITH_FAAD2=OFF WITH_FFMPEG=TRUE WITH_FFMPEG=TRUE WITH_GFX_H264=ON WITH_GPROF=OFF WITH_GSM=OFF WITH_GSSAPI=OFF WITH_ICU=OFF WITH_IPP=OFF WITH_JPEG=OFF WITH_LAME=OFF WITH_LIBRARY_VERSIONING=ON WITH_LIBSYSTEMD=ON WITH_MACAUDIO=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO_AVAILABLE=0 WITH_MANPAGES=ON WITH_MBEDTLS=OFF WITH_OPENCL=OFF WITH_OPENH264=TRUE WITH_OPENH264=TRUE WITH_OPENH264_LOADING=OFF WITH_OPENSLES=OFF WITH_OPENSSL=ON WITH_OSS=ON WITH_PAM=ON WITH_PCSC=ON WITH_PROFILER=OFF WITH_PROXY=ON WITH_PROXY_MODULES=OFF WITH_PULSE=ON WITH_SAMPLE=OFF WITH_SANITIZE_ADDRESS=OFF WITH_SANITIZE_ADDRESS_AVAILABLE=1 WITH_SANITIZE_MEMORY=OFF WITH_SANITIZE_MEMORY_AVAILABLE=1 WITH_SANITIZE_THREAD=OFF WITH_SANITIZE_THREAD_AVAILABLE=1 WITH_SERVER=ON WITH_SERVER_CHANNELS=ON WITH_SERVER_INTERFACE=ON WITH_SHADOW=ON WITH_SMARTCARD_INSPECT=OFF WITH_SOXR=OFF WITH_SSE2=ON WITH_SWSCALE=OFF WITH_THIRD_PARTY=OFF WITH_VAAPI=OFF WITH_VALGRIND_MEMCHECK=OFF WITH_VALGRIND_MEMCHECK_AVAILABLE=1 WITH_WAYLAND=ON WITH_WINPR_TOOLS=ON WITH_X11=ON WITH_X264=OFF WITH_XCURSOR=ON WITH_XDAMAGE=ON WITH_XEXT=ON WITH_XFIXES=ON WITH_XI=ON WITH_XINERAMA=ON WITH_XKBFILE=ON WITH_XRANDR=ON WITH_XRENDER=ON WITH_XSHM=ON WITH_XTEST=ON WITH_XV=ON WITH_ZLIB=ON
Build type:          Release
CFLAGS:              -D_FORTIFY_SOURCE=2 -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -Wall -Wno-unused-result -Wno-unused-but-set-variable -Wno-deprecated-declarations -fvisibility=hidden -Wimplicit-function-declaration -Wredundant-decls -fno-omit-frame-pointer -DWINPR_DLL
Compiler:            GNU, 10.1.0
Target architecture: x64
```

* CMD `xfreerdp /rfx /v:[host] /u:[user] /log-level:trace /sound:sys:alsa /smartcard`
* OS version connecting to (server side): Windows 10

**Environment:**
 - OS: Arch Linux
 - Architecture: amd64

**Comment by akallabeth** [comment]
@Raven24 Best check the following:
1. Does the required smartcard reader feature work locally on your machine? (If not then nothing we can do)
1. You can try with `/client-build-number:2600` to force the smartcard channel to an older feature set (annoying, but at least documented, feature `detection` done by the RDP server)

**Comment by akallabeth** [comment]
@Raven24 another thing you can try (if you have the hardware / software) is if it works at all via RDP (check with the MS Rdp client, if it does not work there it is a protocol limitation)

**Comment by Raven24** [comment]
1. On my local Arch Linux machine, using PCSCd with the cyberjack driver and the (quite outdated) [Linux version](https://www.a-trust.at/de/signaturkarten/middleware/installation/) of the A-Sign client I can list the card, just like on Windows.
![image](https://user-images.githubusercontent.com/117366/89557033-80321c00-d812-11ea-8ebf-b1ae273efaeb.png)

2. Here is the log of the run with `/client-build-number:2600`: 
[freerdp_smartcard_2600.txt](https://github.com/FreeRDP/FreeRDP/files/5036334/freerdp_smartcard_2600.txt)
Unfortunately, that didn't seem to change anything.

3. Using a Windows 10 Client to connect with mstsc.exe I get this in the device manager - looks to forward some more information about OS and Driver verison
![image](https://user-images.githubusercontent.com/117366/89557820-b0c68580-d813-11ea-84ed-662641724b4a.png)
Interesting to see in the WTS (Windows Terminal Server?) tab, with the mstsc client it says it uses "RDP-Tcp#123" and with xfreerdp it says "Console"... what ever that means.
![image](https://user-images.githubusercontent.com/117366/89557951-e23f5100-d813-11ea-901a-74314fc57ece.png)
![image](https://user-images.githubusercontent.com/117366/89558145-26caec80-d814-11ea-9310-f1ca28af61e7.png)

**Comment by Raven24** [comment]
here's also the output of `pcsc_scan` on linux:

```
# pcsc_scan 
Using reader plug'n play mechanism
Scanning present readers...
0: REINER SCT cyberJack pp_a (8108109235) 00 00
 
Thu Aug  6 18:44:28 2020
 Reader 0: REINER SCT cyberJack pp_a (8108109235) 00 00
  Event number: 0
  Card state: Card inserted, 
  ATR: 3B DF 18 00 81 31 FE 58 80 31 B0 52 02 04 64 05 C9 03 AC 73 B7 B1 D4 22

ATR: 3B DF 18 00 81 31 FE 58 80 31 B0 52 02 04 64 05 C9 03 AC 73 B7 B1 D4 22
+ TS = 3B --> Direct Convention
+ T0 = DF, Y(1): 1101, K: 15 (historical bytes)
  TA(1) = 18 --> Fi=372, Di=12, 31 cycles/ETU
    129032 bits/s at 4 MHz, fMax for Fi = 5 MHz => 161290 bits/s
  TC(1) = 00 --> Extra guard time: 0
  TD(1) = 81 --> Y(i+1) = 1000, Protocol T = 1 
-----
  TD(2) = 31 --> Y(i+1) = 0011, Protocol T = 1 
-----
  TA(3) = FE --> IFSC: 254
  TB(3) = 58 --> Block Waiting Integer: 5 - Character Waiting Integer: 8
+ Historical bytes: 80 31 B0 52 02 04 64 05 C9 03 AC 73 B7 B1 D4
  Category indicator byte: 80 (compact TLV data object)
    Tag: 3, len: 1 (card service data byte)
      Card service data byte: B0
        - Application selection: by full DF name
        - BER-TLV data objects available in EF.DIR
        - BER-TLV data objects available in EF.ATR
        - EF.DIR and EF.ATR access services: by GET RECORD(s) command
        - Card with MF
    Tag: 5, len: 2 (card issuer's data)
      Card issuer data: 02 04
    Tag: 6, len: 4 (pre-issuing data)
      Data: 05 C9 03 AC
    Tag: 7, len: 3 (card capabilities)
      Selection methods: B7
        - DF selection by full DF name
        - DF selection by path
        - DF selection by file identifier
        - Short EF identifier supported
        - Record number supported
        - Record identifier supported
      Data coding byte: B1
        - EF of TLV structure supported
        - Behaviour of write functions: proprietary
        - Value 'FF' for the first byte of BER-TLV tag fields: valid
        - Data unit in quartets: 2
      Command chaining, length fields and logical channels: D4
        - Command chaining
        - Extended Lc and Le fields
        - Logical channel number assignment: by the card
        - Maximum number of logical channels: 5
+ TCK = 22 (correct checksum)

Possibly identified card (using /usr/share/pcsc/smartcard_list.txt):
3B DF 18 00 81 31 FE 58 80 31 B0 52 02 04 64 05 C9 03 AC 73 B7 B1 D4 22
        Austrian "e-Card" (=Health Card) of the 4th generation. (HealthCare)
```

**Comment by akallabeth** [comment]
ok, looks like some manipulation of the initial config data sent is done by the ms side of things.
Need to have a look at that when time permits.
Could you check if the `<ATR>` string of local access matches the one with remote access?

**Comment by akallabeth** [comment]
Interesting part is the return for `SCARD_ATTR_ATR_STRING` `SCardGetAttrib` call from pcsc.
Could you check `smartcard_GetAttrib_Call` call return values?

**Comment by Raven24** [comment]
I haven't had the chance to dig deeper into the code at the moment, but I think it might be a problem with my reader device or the cyberjack driver. Will try it again with a different device next week when I'm at the office again.

In the meantime there was something odd in `smartcard_GetAttrib_Call` I noticed:
The [PCSC-lite API docs](https://pcsclite.apdu.fr/api/group__API.html#gaacfec51917255b7a25b94c5104961602) specify to first call `SCardGetAttrib` with a `NULL` buffer so it can write the required size to the reference in the fourth parameter. Then malloc the buffer and call the function again, to write to the correctly sized buffer. Currently in the freerdp code there is only one call to that function(?) - not sure if that's a problem, I didn't follow all the code paths.

**Comment by akallabeth** [comment]
@Raven24 there is `SCARD_AUTOALLOCATE` and the second call is only required if you do not provide a previously allocated result buffer.
If you check `smartcard_GetAttrib_Call` then you see that the buffer is either allocated or the buffer size set to `SCARD_AUTOALLOCATE`

**Comment by Raven24** [comment]
Alright, definitely a driver problem. I tested it again with a HID Omnikey reader (without pinpad) and the smartcard was redirected as expected. 
I then started to poke around in the pcsc-cyberjack source, since you made me curious about the return value of `SCARD_ATTR_ATR_STRING`. Indeed, there was a problem that caused it to return an empty value - so I cobbled together a quick-and-dirty patch, that makes the function return the actual value.
For reference, here is my [post on the manufacturer forum](https://forum.reiner-sct.com/index.php?/topic/4779-scard_attr_atr_string-wird-nichtleer-zur%C3%BCckgegeben/) (german) where I submit my fix as a patch.

Thanks again for helping!
